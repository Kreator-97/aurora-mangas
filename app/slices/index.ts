export { default as uiSlice } from './uiSlice'
export { default as authSlice } from './authSlice'
export { default as shoppingCartSlice } from './shoppingCartSlice'
