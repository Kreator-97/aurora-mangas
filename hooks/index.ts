export * from './useForm'
export * from './useCounter'
