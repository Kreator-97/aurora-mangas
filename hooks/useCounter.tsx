import { useState } from 'react'

interface Props {
  initial?: number;
  minValue?: number;
}

export const useCounter = ({initial = 0, minValue }:Props) => {
  const [ counter, setCounter] = useState(initial)

  const increment = (factor:number) => {
    setCounter(counter + factor)
  }

  const decrement = (factor:number) => {
    if ( minValue !== undefined ) {
      const max = Math.max(minValue, counter-factor)
      setCounter(max)
      return
    }
    setCounter(counter - factor)
  }

  const reset = () => setCounter(initial)

  return { increment, decrement, reset, counter }
}
