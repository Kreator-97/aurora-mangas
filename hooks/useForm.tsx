import { ChangeEvent, useState } from 'react'

export function useForm<T> (initialValue: T) {
  const [ formState, setFormState ] = useState<T>(initialValue)
  
  const onInputChange = ({target}:ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
    const {name, value} = target
    setFormState({
      ...formState,
      [name]: value,
    })
  }

  const onResetForm = () => {
    setFormState(initialValue)
  }

  return {
    ...formState,
    formState,
    onInputChange,
    onResetForm
  }
}
