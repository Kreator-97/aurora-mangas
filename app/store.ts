import { configureStore } from '@reduxjs/toolkit'
import { uiSlice, authSlice, shoppingCartSlice } from './slices'

export const store = configureStore({
  reducer: {
    ui: uiSlice,
    auth: authSlice,
    cart: shoppingCartSlice
  },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
