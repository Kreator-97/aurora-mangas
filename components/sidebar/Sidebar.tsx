import { signOut } from 'next-auth/react'
import { useRouter } from 'next/router'
import { HiX } from 'react-icons/hi'

import { useAppDispatch, useAppSelector } from '../../app/hooks'
import { closeSidebar } from '../../app/slices/uiSlice'

export const Sidebar = () => {
  const { ui, auth } = useAppSelector( (state) => state)
  const dispatch = useAppDispatch()
  const router = useRouter()

  const { user } = auth
  const { isSidebarOpen } = ui
  const closeSideBar = () => {
    dispatch(closeSidebar())
  }

  const navigateTo = (url:string) => {
    closeSideBar()
    router.push(url)
  }

  return (
    <div className={`min-h-screen max-w-[360px] bg-dark fixed top-0 left-0 w-full z-50 p-4 transition-transform ${ isSidebarOpen ? '' : '-translate-x-full'}`}>
      <div className='flex justify-between items-center mb-2'>
        <div className='w-8'></div>
        <h2 className='text-2xl'>{ user?.fullname ? user.fullname : 'Bienvenido' }</h2>
        <HiX
          className='cursor-pointer'
          size={32}
          onClick={ () => closeSideBar() }
          data-testid="close-icon"
        />
      </div>
      {
        auth.isLogged
          ? (<>
            <div className='flex gap-2 mb-4'>
              <button
                className='text-alert w-full'
                onClick={ () => signOut() }
                data-testid="signout"
              >
                Cerrar sesión
              </button>
            </div>
          </>)
          : (<>
            <p className='text-center mb-2'>Ingresa para gestionar tu cuenta</p>
            <div className='flex gap-2 mb-4'>
              <button className='btn bg-accent w-full' onClick={ () => navigateTo('/auth/login') }>Ingresar</button>
              <button className='btn ghost w-full' onClick={() => navigateTo('/auth/register')}>Crear cuenta</button>
            </div>
          </>)
      }
      {
        auth.isLogged && (
          <>
            <section className='border-t border-strokeLight border-solid'>
              <p className='text-xl text-center pt-2'>Sobre tu cuenta</p>
              <button
                className='text-xl py-2 hover:text-accent cursor-pointer block'
                onClick={ () => navigateTo('/orders') }
              >
                Lista de pedidos
              </button>
              <button
                className='text-xl py-2 hover:text-accent cursor-pointer block'
              >
                Datos personales
              </button>
              <button
                className='text-xl py-2 hover:text-accent cursor-pointer block'
                onClick={ () => navigateTo('/subscriptions')}
              >
                Suscripciones
              </button>
            </section>
          </>)
      }
      {
        auth.user?.role === 'ADMIN' && (
          <>
            <section className='border-t border-strokeLight border-solid'>
              <p className='text-xl text-center pt-2'>Administrador</p>
              <button
                className='text-xl py-2 hover:text-accent cursor-pointer block'
                onClick={ () => navigateTo('/admin/orders') }
              >
                Ver pedidos realizados
              </button>
              <button
                className='text-xl py-2 hover:text-accent cursor-pointer block'
                onClick={ () => navigateTo('/admin/series') }
              >
                Series
              </button>
              <button
                className='text-xl py-2 hover:text-accent cursor-pointer block'
                onClick={ () => navigateTo('/admin/mangas') }
              >
                Mangas
              </button>
              <button
                className='text-xl py-2 hover:text-accent cursor-pointer block'
                onClick={ () => navigateTo('/admin/authors') }
              >
                Autores
              </button>

              <button
                className='text-xl py-2 hover:text-accent cursor-pointer block'
                onClick={ () => navigateTo('/admin/analitics/best-sellers') }
              >
                Más vendidos
              </button>
            </section>
          </>
        )
      }
    </div>
  )
}
