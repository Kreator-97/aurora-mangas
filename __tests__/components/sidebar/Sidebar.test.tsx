import { fireEvent, render, screen } from '@testing-library/react'
import { Provider } from 'react-redux'
import { store } from '../../../app/store'
import { Sidebar } from '../../../components/sidebar/Sidebar'
import { useAppDispatch, useAppSelector } from '../../../app/hooks'
import { signOut } from 'next-auth/react'

jest.mock('../../../app/hooks')
jest.mock('next-auth/react')

const dispatch = jest.fn()

describe('tests on Sidebar component', () => {

  test('should to match with snapshot', () => {
    (useAppSelector as jest.Mock).mockReturnValue({
      ui: { isSidebarOpen: true },
      auth: { isLogged: false, user: null },
    })

    const { container } = render(
      <Provider store={store}>
        <Sidebar />
      </Provider>
    )
    expect(container).toMatchSnapshot()
    expect(screen.getByText('Bienvenido')).toBeInTheDocument
    expect(screen.getByText('Ingresa para gestionar tu cuenta')).toBeInTheDocument
    expect(screen.getByText('Ingresar')).toBeInTheDocument
    expect(screen.getByText('Crear cuenta')).toBeInTheDocument
  })
  
  test('should to show username on sidebar when is authenticated', () => {
    const fullname = 'Adrian gomez';
    (useAppSelector as jest.Mock).mockReturnValue({
      ui: { isSidebarOpen: true },
      auth: { isLogged: true, user: { fullname } },
    })
    render(
      <Provider store={store}>
        <Sidebar />
      </Provider>
    )

    expect(screen.getByText(fullname)).toBeInTheDocument
  })

  test('should to dispatch closeSideBar()', () => {

    (useAppSelector as jest.Mock).mockReturnValue({
      ui: { isSidebarOpen: true },
      auth: { isLogged: true, user: null },
    });

    (useAppDispatch as jest.Mock).mockReturnValue(dispatch)
    render(
      <Provider store={store}>
        <Sidebar />
      </Provider>
    )

    const closeIcon = screen.getByTestId('close-icon')
    fireEvent.click(closeIcon)

    expect(dispatch).toHaveBeenCalledWith({
      payload: undefined, type: 'ui/closeSidebar',
    })

  })

  test('should to call signOut()', () => {
    (signOut as jest.Mock).mockReturnValue(() => {})

    render(
      <Provider store={store}>
        <Sidebar />
      </Provider>
    )

    const btnSignOut = screen.getByTestId('signout')
    fireEvent.click( btnSignOut )

    expect(signOut).toHaveBeenCalled()
  })
})
