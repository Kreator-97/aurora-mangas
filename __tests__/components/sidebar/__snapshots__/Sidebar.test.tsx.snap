// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on Sidebar component should to match with snapshot 1`] = `
<div>
  <div
    class="min-h-screen max-w-[360px] bg-dark fixed top-0 left-0 w-full z-50 p-4 transition-transform "
  >
    <div
      class="flex justify-between items-center mb-2"
    >
      <div
        class="w-8"
      />
      <h2
        class="text-2xl"
      >
        Bienvenido
      </h2>
      <svg
        class="cursor-pointer"
        data-testid="close-icon"
        fill="currentColor"
        height="32"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <p
      class="text-center mb-2"
    >
      Ingresa para gestionar tu cuenta
    </p>
    <div
      class="flex gap-2 mb-4"
    >
      <button
        class="btn bg-accent w-full"
      >
        Ingresar
      </button>
      <button
        class="btn ghost w-full"
      >
        Crear cuenta
      </button>
    </div>
    <hr />
    <section>
      <button
        class="text-xl py-2 hover:text-accent cursor-pointer block"
      >
        Listar por géneros
      </button>
      <button
        class="text-xl py-2 hover:text-accent cursor-pointer block"
      >
        Lista de deseos
      </button>
      <button
        class="text-xl py-2 hover:text-accent cursor-pointer block"
      >
        Más vendidos
      </button>
    </section>
  </div>
</div>
`;
